docs/dev/design/new-notifications-system.rst

```diff
@@ -0,0 +1,459 @@
+
+Notifications list
+~~~~~~~~~~~~~~~~~~
+
+.. http:get:: /api/v3/users/(str:user_username)/notifications/
+
+    Retrieve a list of all the notifications for this user.
+
+.. http:get:: /api/v3/projects/(str:project_slug)/notifications/
+
+    Retrieve a list of all the notifications for this project.
+
+.. http:get:: /api/v3/organizations/(str:organization_slug)/notifications/
+
+    Retrieve a list of all the notifications for this organization.
+
+.. http:get:: /api/v3/projects/(str:project_slug)/builds/(int:build_id)/notifications/
+
+    Retrieve a list of all the notifications for this build.
+
+    **Example response**:
+
+    .. sourcecode:: json
+
+        {
+            "count": 25,
+            "next": "/api/v3/projects/pip/builds/12345/notifications/?unread=true&sort=type&limit=10&offset=10",
+            "previous": null,
+            "results": [
+                {
+                    "slug": "cancelled-by-user",
+                    "state": "unread",
+                    "dismissable": false,
+                    "bell": false,
+                    "attached_to": "build",
+                    "message": {
+                        "header": "User action",
+                        "body": "Build cancelled by the user.",
+                        "type": "error",
+                        "icon": "fa-exclamation",
+                        "color": "red"
+                    }
+                }
+            ]
+        }
+
+    :query boolean unread: return only unread notifications
+    :query string type: filter notifications by type (``error``, ``note``, ``tip``)
+    :query string sort: sort the notifications (``type``, ``date`` (default))
+
+
+Notification create
+~~~~~~~~~~~~~~~~~~~
+
+
+.. http:post:: /api/v3/projects/(str:project_slug)/builds/(int:build_id)/notifications/
+
+    Create a notification for the resource.
+    In this example, for a ``Build`` resource.
+
+    **Example request**:
+
+    .. sourcecode:: json
+
+        {
+            "slug": "cancelled-by-user",
+            "type": "error",
+            "state": "unread",
+            "dismissable": false,
+            "bell": false,
+        }
+
+
+.. note::
+
+   Similar API endpoints will be created for each of the resources
+   we want to attach a ``Notification`` (e.g. ``User``, ``Organization``, etc)
+
+
+Backward compatibility
+----------------------
+
+It's not strickly required, but if we want, we could extract the current notification logic from:
+
+* Django templates
+
+  * "Don't want ``setup.py`` called?"
+  * ``build.image`` config key is deprecated
+  * Configuration file is required
+  * ``build.commands`` is a beta feature
+
+* ``Build.error`` fields
+
+  * Build cancelled by user
+  * Unknown exception
+  * ``build.os`` is not found
+  * No config file
+  * No checkout revision
+  * Failed when cloning the repository
+  * etc
+
+and iterate over all the ``Build`` objects to create a ``Notification`` object for each of them.
```

Comment:
Are there notifications in particular you're thinking about here? If not, this seems like a step we could probably skip. Old builds aren't very valuable already and already have a `Build.error`, so it would just be a matter of conditionally using `Notification` or `Build.error` to gather messages (notifications stored in templates will be discarded entirely though).

Reply:
The other point of this "backward compatibility" is about migrating the "old build notifications" to the new `Notification` model we are defining here. I wouldn't start here either as first step without doing some testing on production.

Old builds will keep showing old notifications based on `Build.error` fields. New builds will show only notifications from the new system.